### PR TITLE
50 implement image editing

### PR DESCRIPTION
**Pull request by shachar-langer**

### CODE REVIEW

File: scss/popup.scss [hunk 1]
```diff
@@ -1,10 +1,12 @@
 @use '../node_modules/draft-js/dist/Draft.css';
 @use '../node_modules/emoji-mart/css/emoji-mart.css';
+@use './components/screenshots.scss';
```
Comment (will-weiss) [inline scss/popup.scss:3]: 👍

File: src/background/responders.ts [hunk 2]
```diff
@@ -1,4 +1,5 @@
 import { EditorState } from 'draft-js'
+import { pullAt } from 'lodash'
```
Comment (will-weiss) [inline src/background/responders.ts:2]: did we use this?

File: src/background/screenshot.ts [hunk 3]
```diff
@@ -41,6 +41,6 @@ export async function takeScreenshot(tab: TabInfo, tabs: typeof browser.tabs, di
       },
     })
   } catch (error) {
-    dispatchBackgroundActions.screenshotCaptureFailure(error)
+    dispatchBackgroundActions.screenshotCaptureFailure({ error })
```
Comment (will-weiss) [inline src/background/screenshot.ts:44]: 👍

File: src/popup/app.tsx [hunk 4]
```diff
@@ -2,14 +2,16 @@ import * as React from 'react'
...
-import { activeTab } from '../selectors'
+import { activeTab, deleteScreenshotDisabled, takeScreenshotDisabled } from '../selectors'
 
 type AppProps = {
   state: AppState
   dispatchUserActions: Dispatchers<UserAction>
 }
 
 export function App({ state, dispatchUserActions }: AppProps): null | JSX.Element {
+  console.log('state.auth.state', state.auth.state)
```
Comment (will-weiss) [inline src/popup/app.tsx:13]: rm

File: src/popup/components/action-button.tsx [hunk 5]
```diff
@@ -3,9 +3,11 @@ import * as svgs from './action-svgs'
 
 type ActionButtonKind = keyof typeof svgs
 
-export function ActionButton({ kind, onClick }: { kind: ActionButtonKind; onClick(): void }) {
+type ActionButtonProps = { kind: ActionButtonKind; onClick(): void; disabled?: boolean }
+
+export function ActionButton({ kind, onClick, disabled }: ActionButtonProps): JSX.Element {
   return (
-    <button className="svg-btn" onClick={onClick}>
+    <button className={`svg-btn ${kind}`} onClick={onClick} disabled={disabled}>
```
Comment (will-weiss) [inline src/popup/components/action-button.tsx:10]: 👍

File: src/popup/components/screenshots.tsx [hunk 6]
```diff
@@ -3,29 +3,69 @@ import * as React from 'react'
...
 }
 
-export function ScreenshotThumbnail({ screenshot, startEditingScreenshot }: ScreenshotThumbnailProps) {
+function CloseButton({ onClick }: { onClick: () => void }): JSX.Element {
   return (
-    <>
-      <img className="screenshot-thumbnail" src={screenshot.uri} />
-      <button onClick={startEditingScreenshot}>Edit</button>
-    </>
+    <button className="close-button" onClick={onClick}>
```
Comment (will-weiss) [inline src/popup/components/screenshots.tsx:19]: Can we make this into an `svg-btn` somehow?

File: src/tests/happy-path.test.ts [hunk 7]
```diff
@@ -173,10 +173,39 @@ describe('happy path', () => {
...
+      const screenshotImage = screenshotThumbnail?.querySelector('.screenshot-image')
       const screenshotUri = activeTab(getState())?.feedbackState.screenshots[0].uri
-      expect(screenshotThumbnail).to.have.property('src', screenshotUri)
+      expect(screenshotImage).to.have.property('src', screenshotUri)
+    })
+
+    it('takes a screenshot', async () => {
+      const tab = activeTab(getState())!
+      expect(tab.feedbackState.screenshots).to.have.lengthOf(1)
+      const takeScreenshotButton = app().querySelector('.TakeSnapshot')! as HTMLButtonElement
```
Comment (will-weiss) [inline src/tests/happy-path.test.ts:185]: Let's rename `TakeSnapshot` in the codebase to be `TakeScreenshot`

File: src/tests/happy-path.test.ts [hunk 8]
```diff
@@ -173,10 +173,39 @@ describe('happy path', () => {
...
     it('renders the screenshot', () => {
       const screenshotThumbnail = app().querySelector('.twitter-interface > .screenshots > .screenshot-thumbnail')
+
+      const screenshotImage = screenshotThumbnail?.querySelector('.screenshot-image')
       const screenshotUri = activeTab(getState())?.feedbackState.screenshots[0].uri
-      expect(screenshotThumbnail).to.have.property('src', screenshotUri)
+      expect(screenshotImage).to.have.property('src', screenshotUri)
+    })
+
+    it('takes a screenshot', async () => {
```
Comment (will-weiss) [inline src/tests/happy-path.test.ts:182]: `it('takes a screenshot when the TakeScreenshot button is clicked'`

File: src/tests/happy-path.test.ts [hunk 9]
```diff
@@ -173,10 +173,39 @@ describe('happy path', () => {
...
+      const tab = activeTab(getState())!
+      expect(tab.feedbackState.screenshots).to.have.lengthOf(1)
+      const takeScreenshotButton = app().querySelector('.TakeSnapshot')! as HTMLButtonElement
+      takeScreenshotButton.click()
+      await whenState(mocks.backgroundWindow.store, state => ensureActiveTab(state).feedbackState.screenshots.length === 2)
+      const screenshotImages = app().querySelectorAll('.screenshot-image')
+      expect(screenshotImages).to.have.lengthOf(2)
     })
 
+    it('delete a screenshot', async () => {
```
Comment (will-weiss) [inline src/tests/happy-path.test.ts:192]: `it('deletes a screenshot when the close-button is clicked'

File: src/tests/happy-path.test.ts [hunk 10]
```diff
@@ -173,10 +173,39 @@ describe('happy path', () => {
 
     it('renders the screenshot', () => {
       const screenshotThumbnail = app().querySelector('.twitter-interface > .screenshots > .screenshot-thumbnail')
+
+      const screenshotImage = screenshotThumbnail?.querySelector('.screenshot-image')
       const screenshotUri = activeTab(getState())?.feedbackState.screenshots[0].uri
-      expect(screenshotThumbnail).to.have.property('src', screenshotUri)
+      expect(screenshotImage).to.have.property('src', screenshotUri)
```
Comment (will-weiss) [inline src/tests/happy-path.test.ts:179]: Test that there's no close-button here

File: src/tests/happy-path.test.ts [hunk 11]
```diff
@@ -173,10 +173,39 @@ describe('happy path', () => {
...
+    })
+
+    it('takes a screenshot', async () => {
+      const tab = activeTab(getState())!
+      expect(tab.feedbackState.screenshots).to.have.lengthOf(1)
```
Comment (will-weiss) [inline src/tests/happy-path.test.ts:184]: We can test the disabled logic like so

```typescript
let screenshotsLength: number = 1
while (screenshotsLength < 9) {
      const takeScreenshotButton = app().querySelector('.TakeScreenshot')! as HTMLButtonElement
      takeScreenshotButton.click()
      await whenState(mocks.backgroundWindow.store, state => ensureActiveTab(state).feedbackState.screenshots.length === screenshotsLength + 1)
      const screenshotImages = app().querySelectorAll('.screenshot-image')
      expect(screenshotImages).to.have.lengthOf(screenshotsLength + 1)
      screenshotsLength++
}

expect(app().querySelector('.TakeScreenshot')).to.have.property('disabled', true)
```

**Review by will-weiss — CHANGES_REQUESTED** [review]

File: src/popup/components/screenshots.tsx (hunk 6 above)
Comment (shachar-langer) [reply src/popup/components/screenshots.tsx:19]: I can export the SVG from Figma but why?

**Review by will-weiss — APPROVED** [review]

**Pull request merged**